### PR TITLE
Improve the Reusable Blocks UI by relying on multi entity save flow.

### LINKED ISSUE #11784: Save reusable blocks on "Save"
Saving a post/page should also first save edited reusable blocks. 

To reproduce:
1. On a page/post, edit a reusable block.
2. Forget to hit "save" on the reusable block. 
3. Save the page/post.
4. Note that the reusable block edit is lost.

It seems expected behavior is that the edited reusable block would be automatically saved in the page/post Save. 

Using 4.3.0.

### LINKED ISSUE #12079: Reusable Blocks: Incorrect label appears in block breadcrumb component
When I hover over a reusable block before I click on it shows the label in the top right corner to be "Reusable Block" as expected. When I'm editing the reusable block and hover over a block inside the reusable block the label in the top right corner shows "Reusable Template" as the parent when I would expect "Reusable Block". I'm a site developer working to understand exactly how Gutenberg is put together and works. When using Gutenberg I've started to use the hover labels in the top right to tell me where I am and is incredibly helpful when working with block nesting (inner blocks). If this is intentional and part of how reusable blocks work could you link me to the documentation? If this is only a labeling mistake, would you mind mending it? It would help with clarity for me during this change of editor.

Thanks for your work!

## To Reproduce
1. Create a post using Gutenberg.
2. Insert a reusable block.
3. Hover over the reusable block and observe the label in the top right corner.
4. Click on the reusable block.
5. Click on the "edit" button in the top right.
6. Hover over one of the blocks inside the reusable block and observe the label in the top right corner.

## Expected behavior
I was expecting the labels to both use "Reusable Block".

## Screenshots
48733424-49643e80-ec08-11e8-9225-e840897d3cd6
48733431-4bc69880-ec08-11e8-9932-58c56b5a535b

## Desktop
 - Gutenberg: v4.4.0
 - OS: macOS v10.13.5
 - Browser: Firefox
 - Version: 63.0.3

### LINKED ISSUE #12255: Dragging a reusable multi column block into itself, the block is messed up or removed.
When I drag a reusable block with a multi column block into itself, the block is messes up or removed.
This issue is similar to https://github.com/WordPress/gutenberg/issues/9303
but it's not for a reusable block.

**To Reproduce**
1. Insert a multi column block.
2. Put any blocks in each column.
3. Register the multi column block to a reusable block.
4. Drag the block into itself.

**Expected behavior**
When a multi column block is in a reusable block, it should be locked and not accept dragging.

**Screenshots**
b234deccc35b30ced3f11fbf9f324666

8cf9e0a37c8dc836350351948807379f

**Desktop:**
 - OS: Windows 10
 - Browser: Chrome 70.0.3538.102, Firefox 63.0.3 

Gutenberg version 4.5.1 and the Test Gutenberg web site.

### LINKED ISSUE #16614: Custom block update rendering when reused
I developed a custom (Gutenberg) block that have to update is content on almost every block manipulation on the block editor.

It works great, mostly: i "listen" to changes with a `wp.data.subscribe` function and update the block attributes with `wp.data.dispatch( 'core/block-editor' ).updateBlockAttributes`.

The problem is that the blocks stop updating when converted to a reusable block or put in one (like a reusable group). It's fine in the reusable block manager, but the feature that let you edit a reusable block inside the post it is used on is like a frame that don't fire my wp.data.subscribe function.

How do i check for modifications in this "window" and update blocks inside it?

It's just a backend problem. The client side rendering is dynamic so it works fine.

I've searched a lot, and asked here ( where you can also find the code) : https://wordpress.stackexchange.com/questions/342753/custom-block-update-rendering-when-reused.

I'm an experienced WP developer, but pretty new to React so sorry if my vocabulary is off...

Thank you,

### LINKED ISSUE #18560: Sidebar doesn't appear when selecting a block inside a Reusable Block
**Describe the bug**
Before the update, when you select a block inside the reusable block, it's settings would show in the sidebar under the "Block" tab. Now the tab is empty. This issue isn't there is a block isn't reusable, but a regular block.

**To reproduce**
1. Create 2 group blocks with paragraphs in them
2. Make one block reusable.
3. Select block inside regular group and see the settings of the block in the sidebar.
4. Click "edit" on the reusable group.
5. Select the paragraph block
6. See empty sidebar on the right or reusable group settings (don't know on what this depends, but it's either way and isn't consistent).

**Expected behavior**
The inner block of the reusable block have settings on the right sidebar when selected

**Screenshots**
[block settings in a simple group](http://i.imgur.com/h8kaVu4.png)
[block settings in a reusable group](http://i.imgur.com/W6zZtbY.png)
[block settings in a reusable group v2](http://i.imgur.com/ne6tZlT.png)

**Desktop (please complete the following information):**
 - OS: Ubuntu 18.04
 - Browser: Chrome 78
 - Version : WP 5.3

### LINKED ISSUE #19372: Block Navigation doesn't display contents of a Reusable Block
**Describe the bug**
The blocks contained inside a Reusable Block aren't shown in the Block Navigation in a sub-tree similar to how other parent blocks such as Columns show the blocks they contain.

**To reproduce**
1. Insert multiple blocks in the Block Editor.
2. Highlight multiple blocks and select Convert to Reusable Block.
3. Name and Save the Reusable Block.
4. Open the Block Navigation from the Editor Toolbar and only Reusable Block is shown, not the contained blocks as a sub-tree.

**Expected behavior**
The Block Navigation should ideally reveal the contents of a Reusable Block in a sub-tree similar to how Columns, Groups, Media & Text and other blocks that contain children display a sub-tree when highlighted.

**Additional context**
- WordPress: Version 5.3.3-alpha-46995

### LINKED ISSUE #19436: Reusable block selection issues
**Describe the bug**

I'm not sure if any of these issues have been reported before.

* It's currently possible to have a block inside a reusable block selected while another outside of it is selected. This seems really confusing to me.
* The rich text toolbar buttons are duplicated, probably because they're all going to the first slot. (The slot name is global.)

Screenshot 2020-01-06 at 17 46 57

* Pressing `Tab` from a block inside a reusable block focusses the next block outside of the reusable block instead of the inspector. This seems to be caused by having `WritingFlow` inside reusable blocks. I'm not sure why this is needed.
* Selection is not cleared after `Cmd+A` and then selecting a block inside a reusable block.

### LINKED ISSUE #22405: Reusable Blocks respond erratically when external media files are dragged & dropped into them
**Describe the bug**
It is currently possible to drag and drop a file from your computer into an existing reusable block. From the editor, it will look like the file was added, but it will not be saved and in some instances it will make the editor crash.

**To reproduce**
1. Create a new post, add a reusable block that has at least two blocks in it.
2. Drag & drop an image from your computer inside the reusable block
3. Click on preview > The newly added image will not be inside the reusable block 
4. Click on "edit reusable block" > The reusable block will become editable but the image will disappear

**Expected behavior**
The drag & drop action should either edit the reusable block directly (then add the image to it and change its state to "being edited") or prevent the user from dropping a file inside at all.

**Screenshots**
With the Core version of Gutenberg (5.4.1) the image will just disappear if you reload the editor, try to preview it or edit the reusable block:
ezgif-4-9ee65ac962a1

With the plugin version of Gutenberg (8.1.0) this action will cause **the editor to crash**:
ezgif-4-bb98b22cd1c6

**Editor version**
- WordPress version: 5.4.1
- The error occurs with and without the plugin, but varies slightly (see above)
- If the Gutenberg plugin is installed, which version is it? 8.1.0

**Desktop**
MacOS 10.13.6 (17G2307)
Google Chrome 79.0.3945.88

### LINKED ISSUE #22832: The toolbar behavior in the "Reusable Blocks" is buggy.
When editing "Reusable Blocks", if you focus on "Reusable Blocks" itself after focusing on the blocks inside, the "block toolbar" will jump to "Reusable Blocks".

**Screenshots**
the_bug

**Editor version**
- WordPress version: 5.4.1
- Gutenberg plugin version : 8.2.1

Note: It also occurs in WordPress 5.4.1 without Gutenberg plugin.

**Desktop (please complete the following information):**
 - OS: macOS Catalina
 - Browser : Chrome, Safari

**Additional context**
When there are multiple blocks inside, I can edit it by somehow moving the focus to another block.

### LINKED ISSUE #26670: Selecting a block inside a reusable block shows "No block selected" message in the sidebar
**Describe the bug**
If you click on a block that's sitting inside a reusable block, no options for that block will show, instead a "**_No block selected_**" message is seen in the sidebar.

**To reproduce**
1. Create a new post and add one or more blocks
2. Convert the blocks into a reusable block
3. Edit the reusable block and select one of the blocks sitting inside.
4. No settings for that block will be described, only a "_**No block selected**_" message.

**Expected behavior**
I would expect to be able to modify a block's settings, even if it is sitting inside a reusable block. 

**Screenshots**
Screen Shot 2020-11-03 at 20 10 11

The error goes away if we edit the reusable block via the "Manage All Reusable Blocks" section:

Screen Shot 2020-11-03 at 20 11 37

**Editor version (please complete the following information):**
- WordPress 5.5.3
- Gutenberg 9.2.2

The problem seems to occur with and without the plugin.

**Desktop (please complete the following information):**
 - Mac OS 10.13.6
 - Chrome 86.0.4240.111

### LINKED ISSUE #26824: Missing/Broken Top Toolbar on Reusable Blocks
**Describe the bug**
Activating the "Top Toolbar" option while editing a reusable block will make the block toolbar disappear. Or, to be more exact, it does appear at the top but seems to be broken or obscured by other UI elements.

**To reproduce**
1. Create a new post and insert a reusable block
2. Change the editor preferece to "Top Toolber"
3. The block Toolbar will appear broken (no incons) ant the very top.

**Expected behavior**
The Top Toolbar option should work on reusable block the same way it works on any other block, and show at the very top.

**Screenshots**
Kapture 2020-11-09 at 13 08 11

The Top Toolbar shows only the "three dot" menu for the selected block, but not the movers or any other buttons. 
Screen Shot on 2020-11-09 at 13:05:10

**Editor version (please complete the following information):**
- WordPress version 5.5.3
- I noticed this on a WordPress.com site (9.3.0) but was able to reproduce it on a self-hosted site, both without and with the Gutenberg plugin (v 9.3.0)

**Desktop (please complete the following information):**
 - Mac OS 10.13.6
 - Chrome  86.0.4240.80

### PR DESCRIPTION
**Pull request by youknowriad**
closes #18560
closes #22832
closes  #26824
closes #26670
closes #19436 
closes #16614 
closes #12255 
closes #12079 
closes #11784 
closes #22405 
closes #19372 

This PR refactors the reusable block to use a controlled innerBlocks. This means several things:

 - It fixes some issues when editing reusable blocks: for instance the inspector sidebar is now visible for selected blocks inside the reusable blocks.
 - We can do things like drag a paragraph out of a reusable block
 - It also means no "edit/save" button on  the reusable block itself, saving is deferred to the "multi entity" saving behavior.

For an initial version, I just show the  "title input" when the block is selected and hide it otherwise but this creates "jumps" in the UI, we need better design here.

### CODE REVIEW

**Comment by mtias** [comment]
This is great, so much code cleanup! Let's see what design improvements can be made and shared between template parts and reusable blocks.

File: packages/block-library/src/block/edit.js [hunk 1]
```diff
@@ -161,23 +114,15 @@ export default function ReusableBlockEdit( {
...
-				{ element }
+				{ isSelected ||
+					( hasInnerBlockSelected && (
+						<ReusableBlockEditPanel
+							title={ title }
+							onChange={ setTitle }
+						/>
+					) ) }
+
+				{ <div { ...innerBlocksProps } /> }
```
Comment (mtias) [inline packages/block-library/src/block/edit.js:125]: Is this div necessary?

Comment (youknowriad) [reply packages/block-library/src/block/edit.js:125]: yes, that's the actual inner blocks.

**Comment by mtias** [comment]
We should disable all transforms for reusable blocks as they crash the editor when you use them and don't really make sense:

image

**Comment by mtias** [comment]
image

This and the fact the inspector now works for individual blocks contained within is amazing.

There are some things that pop up design wise as worth doing, but we can follow up on them afterwards — like indicating when a reusable block has had changes in the reusable block toolbar (has unsaved changes).

We should also move the reusable block name next to the block type icon (like we are doing for template parts) and update the name of the block in things like breadcrumb and block navigation to use the actual title of the block.

**Comment by mtias** [comment]
That list of "closes" is epic

**Comment by chrisvanpatten** [comment]
Fantastic to see; thanks for tackling @youknowriad!

**Comment by Soean** [comment]
Nice PR 👍 

The Edit/Save button is not aligned correctly.

**Before**
Bildschirmfoto 2020-12-29 um 19 14 48

**After**
Bildschirmfoto 2020-12-29 um 19 12 50

**Comment by Soean** [comment]
Nice PR 👍 

I can't duplicate the block or edit it, if a reusable block is used twice in a post.

https://user-images.githubusercontent.com/695201/103306669-6e922080-4a0e-11eb-859f-6572bdf791da.mov

**Comment by youknowriad** [comment]
@Soean This is exactly the bug that was supposed to be fixed in #22639 

It turns out that the branch didn't include the fix. It should be fine now (after a rebase)

**Comment by mtias** [comment]
Without the in-canvas name this feels alright to move forwards and we can follow up for a better presentation in the toolbar of the name. There's something odd in the flow of creating these when the name is set to undefined because you don't get a chance to modify it before creation. It's probably time to add a modal requiring you to input a name before a reusable block is created.

**Review by mtias — APPROVED** [review]

**Comment by Copons** [comment]
We are having increased reports of broken reusable blocks (and posts using them) on WordPress.com, apparently started after this change (see [original issue](https://github.com/Automattic/wp-calypso/issues/49007)).

The issue is technically caused by reusable blocks referencing themselves, which if I'm not mistaken, is currently being addressed in https://github.com/WordPress/gutenberg/pull/27012 (which is not limited to reusable blocks, but anything where such infinite loop could happen).

Regardless, the infinite loop was possible even before this PR.
We suspect that the increased occurrences are caused by the UI changes introduced here.

Before: users had to explicitly edit (and save) a reusable block directly in the block UI itself.
Now: the reusable block is freely editable, and it's then handled by the multi-entity save panel.

What seems to happen is that folks might want to add multiple instances of the same reusable block "in sequence", but they accidentally get "stuck" inside the reusable block group.
E.g.
- Insert reusable block.
- Press enter to "leave" the reusable block group.
- You are unkowingly still inside the reusable block.
- Insert the same reusable block again.
- You might expect it's being added _after_ the previous one, but instead it's _inside_ it.
- Infinite loop!

---

As I said, the infinite loop issue is already being addressed (although, we might probably want to expedite the fix!), but we might also want to reconsider if this very light UI is the right approach for reusable blocks, or any kind of nested blocks.

I consider myself an expert user so I know where to look to figure out the nesting situation.
Still I'd really love to have a at-a-glance, contextual way to understand what I'm editing, rather than having to check the breadcrumbs, outline, or the little parent icon that pops out from the toolbar.

cc @mtias

**Comment by mtias** [comment]
See #27890 for some design considerations and possible improvements

**Comment by paaljoachim** [comment]
The new save flow for reusable blocks creates UX difficulties. Moving the Reusable save away from the block creates more of a challenge to save an update of the block. Check out some of the feedback: [The redesign of reusable blocks that happened in Gutenberg a few weeks back is confusing and makes them much harder to use.](https://github.com/WordPress/gutenberg/issues/28931#issuecomment-781665182) and https://github.com/WordPress/gutenberg/issues/29178#issuecomment-783139205

It would be nice to make the save flow for reusable blocks easier before releasing WordPress 5.7.
@kellychoffman @hedgefield @shaunandrews

**Comment by mtias** [comment]
@paaljoachim there are related tasks in https://github.com/WordPress/gutenberg/issues/27890

**Pull request merged**